src/twister2/builder/smart_builder_plugin.py

```diff
@@ -0,0 +1,109 @@
+
+    It allows to skip building when it was already built for another test.
+    """
+    _lock: FileLock = FileLock(_TMP_DIR + '/smart_builder.lock', timeout=1)
+
+    def __init__(self, output_dir: str | Path) -> None:
+        self._status_file: Path = Path(output_dir) / 'twister_builder.json'
+        self.wait_timeout: int = 10 * 60  # seconds
+
+    def clear(self):
+        logger.info('Cleaning builder status file: %s', self._status_file)
+        with self._lock:
+            self._write_data({})
+
+    def get_status(self, build_dir: str) -> int:
+        with self._lock:
+            data = self._read_data()
+            return data.get(str(build_dir), BuildStatus.NOT_DONE)
+
+    def _read_data(self):
+        with self._status_file.open(encoding='UTF-8') as file:
+            data: dict = json.load(file)
+        return data
+
+    def update_status(self, build_dir: str, status: int) -> int:
+        with self._lock:
+            data = self._read_data()
+            data[str(build_dir)] = status
+            self._write_data(data)
+            return status
+
+    def _write_data(self, data):
+        with self._status_file.open('w', encoding='UTF-8') as file:
+            json.dump(data, file, indent=2)
+
+    def build(
+            self,
+            builder: BuilderAbstract,
+            build_dir: str | Path,
+            platform: str,
+            scenario: str,
+            cmake_args: list[str]
+    ):
+        status = self.get_status(build_dir)
+        if status == BuildStatus.DONE:
+            logger.info('Already build in %s', build_dir)
+            return
+        if status == BuildStatus.NOT_DONE:
+            self.update_status(build_dir, BuildStatus.IN_PROGRESS)
+            try:
+                builder.build(
+                    build_dir=build_dir,
+                    platform=platform,
+                    scenario=scenario,
+                    cmake_args=cmake_args
+                )
+            except Exception:
+                self.update_status(build_dir, BuildStatus.FAILED)
+                raise
+            else:
+                self.update_status(build_dir, BuildStatus.DONE)
+            return
+        if status == BuildStatus.IN_PROGRESS:
+            logger.debug('Waiting for finishing build: %s', build_dir)
+            timeout = time.time() + self.wait_timeout
+            while status == BuildStatus.IN_PROGRESS:
+                time.sleep(1)
+                status = self.get_status(build_dir)
+                if time.time() > timeout:
+                    pytest.fail('Timeout appeared for waiting for build: %s', build_dir)
+            if status != BuildStatus.DONE:
+                pytest.fail('Expected build status %s, but was %s', BuildStatus.DONE, build_dir)
+            return
+        if status == BuildStatus.FAILED:
+            pytest.fail('Building failed: %s', build_dir)
+
+    def pytest_sessionstart(self, session) -> None:
+        if hasattr(session.config, 'workerinput'):
```

Comment:
What do you think about creating in `src/twister2/helper.py` function like `is_xdist_worker`, like:
```python
def is_xdist_worker(session):
    if hasattr(session.config, 'workerinput'):
        return True
    else:
        return False
```
This verification is used quite often in whole Twister v2 project and I think that for people who don't know how `xdist` works internally, this checking `if hasattr(session.config, 'workerinput'):` can be quite mysterious.